components/eamxx/src/share/util/scream_time_stamp.hpp

```diff
@@ -86,6 +86,86 @@ TimeStamp operator- (const TimeStamp& ts, const int dt);
 // If input string is not of the format YYYY-MM-DD-XXXXX, returns an invalid time stamp
 TimeStamp str_to_time_stamp (const std::string& s);
 
+// An enum describing two ways to look at timestamps:
+//  - Linear: treat them as part of a 1d line
+//  - YearlyPerioci: treat them as part of a yearly periodic orbit
+// This is used in the TimeInterval class below to correctly handle time stamps differences
+enum class TimeLine {
+  YearlyPeriodic,
+  Linear
+};
+
+/*
+ * Small struct to deal with time intervals
+ *
+ * The struct simply contains timestamps for [begin,end] interval,
+ * and allows two things: compute the interval length (in days), and check if
+ * a timestamp lies within the interval.
+ *
+ * When te TimeLine arg to the ctor is YearlyPeriodic, the year part of beg/end
+ * time points is ignored. In this case, the length of the time interval is bound
+ * to be in the interval [0,365] (in non-leap years)
+ */
+struct TimeInterval {
+
+  TimeStamp beg;
+  TimeStamp end;
+  TimeLine  timeline = TimeLine::Linear;
+  double length  = -1; // the interval length
+
+  TimeInterval () = default;
+  TimeInterval (const util::TimeStamp& b, const util::TimeStamp& e, TimeLine tl, bool do_compute_length = true)
+   : beg (b), end (e), timeline (tl)
+  {
+    if (do_compute_length)
+      compute_length ();
+  }
+
+  bool contains (const util::TimeStamp& t) const {
+    if (timeline==TimeLine::Linear) {
+      // Compare the full time stamps
+      return beg<=t and t<=end;
+    } else {
+      // Compare the fraction of year for beg/end and t.
+      // Pay extra attention to the case where new year's eve
+      // is in [bec,end]
+      auto t_frac = t.frac_of_year_in_days();
+      auto end_frac = end.frac_of_year_in_days();
+      auto beg_frac = beg.frac_of_year_in_days();
+      bool across_nye = beg.get_month()>end.get_month();
+      if (not across_nye) {
+        return beg_frac<=t_frac and t_frac<=end_frac;
+      } else {
+        // We are either PAST beg or BEFORE end (but not both)
+        return t_frac>=beg_frac or t_frac<=end_frac;
+      }
+    }
+  }
+
+  void compute_length () {
+    bool across_nye = beg.get_month()>end.get_month();
```

Comment:
We only need `across_nye` in the else-block, I think it's easier to read if it's defined there.